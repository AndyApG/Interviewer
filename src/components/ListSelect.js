import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Button } from '../ui/components';
import { CardList } from '.';
import { NodePK } from '../ducks/modules/network';

class ListSelect extends Component {
  constructor(props) {
    super(props);

    this.state = {
      ascending: this.props.initialSortDirection === 'asc',
      filterValue: '',
      property: this.props.initialSortOrder || this.props.labelKey,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.initialSortOrder !== this.props.initialSortOrder) {
      this.setState({
        property: nextProps.initialSortOrder,
      });
    }
    if (nextProps.initialSortDirection !== this.props.initialSortDirection) {
      this.setState({
        ascending: nextProps.initialSortDirection === 'asc',
      });
    }
  }

  /**
    *
    */
  onFilterChange = (event) => {
    this.setState({
      filterValue: event.target.value,
    });
  };

  /**
    * @param {string} current property
    * @return character to indicate sort direction (if applicable)
    */
  getDirection = (property) => {
    if (property === this.state.property) {
      return this.state.ascending ? ' \u25B2' : ' \u25BC';
    }
    return '';
  };

  /**
    * @return sorted list
    */
  getSortedList = () =>
    this.props.nodes.sort(this.compare(this.state.property, this.state.ascending));

  /**
    * @return filtered list
    */
  getFilteredList = list => list.filter(node => (
    this.props.label(node) && this.props.details(node) &&
    (this.props.label(node).toLowerCase().includes(this.state.filterValue.toLowerCase()) ||
    this.props.details(node).some(detail => Object.values(detail).some(
      item => item.toLowerCase().includes(this.state.filterValue.toLowerCase()),
    )))
  ));

  /**
    * @param property to sort by
    */
  setSortBy = (property) => {
    if (this.state.property === property) {
      this.toggleSortDirection();
    } else {
      this.setState({
        ascending: true,
        property,
      });
    }
  };

  /**
    * @param {object} node
    */
  selected = node => !!this.props.selectedNodes.find(current => current[NodePK] === node[NodePK]);

  /**
    * @param property to sort by
    * @param {boolean} ascending
    */
  compare = (property, ascending = true) => {
    let sortOrder = 1;
    if (!ascending) {
      sortOrder = -1;
    }

    return (a, b) => {
      if (!Object.hasOwnProperty.call(a, property) || !Object.hasOwnProperty.call(b, property)) {
        return 0;
      }

      let result = 0;

      const varA = (typeof a[property] === 'string') ?
        a[property].toUpperCase() : a[property];
      const varB = (typeof b[property] === 'string') ?
        b[property].toUpperCase() : b[property];

      if (varA < varB) {
        result = -1;
      } else if (varA > varB) {
        result = 1;
      }
      return result * sortOrder;
    };
  };

  /**
    * changes direction of current sort
    */
  toggleSortDirection = () => {
    this.setState({
      ascending: !this.state.ascending,
    });
  };

  /**
    * toggle whether the card is selected or not.
    * @param {object} node
    */
  toggleCard = (node) => {
    const matchingPK = n => n[NodePK] === node[NodePK];
    const index = this.props.selectedNodes.findIndex(matchingPK);
    if (index !== -1) {
      this.props.onRemoveNode(this.props.nodes.find(matchingPK));
    } else {
      this.props.onSubmitNode(this.props.nodes.find(matchingPK));
    }
  };

  render() {
    const {
      details,
      label,
      sortFields,
    } = this.props;

    return (
      <div className="list-select">
        <div className="list-select__sort">
          { sortFields && sortFields.map(sortField => (
            <Button
              color={this.state.property === sortField.variable ? 'primary' : 'white'}
              key={sortField.variable}
              onClick={() => this.setSortBy(sortField.variable)}
              size="small"
            >
              {(sortField.label || sortField.variable) + this.getDirection(sortField.variable)}
            </Button>
          ))}
          <input type="search" placeholder="Filter" onChange={this.onFilterChange} value={this.state.filterValue} />
        </div>
        <CardList
          details={details}
          label={label}
          nodes={this.getFilteredList(this.getSortedList())}
          onToggleCard={this.toggleCard}
          selected={this.selected}
        />
      </div>
    );
  }
}

ListSelect.propTypes = {
  details: PropTypes.func,
  initialSortOrder: PropTypes.string,
  initialSortDirection: PropTypes.string,
  label: PropTypes.func,
  labelKey: PropTypes.string,
  nodes: PropTypes.array.isRequired,
  onRemoveNode: PropTypes.func,
  onSubmitNode: PropTypes.func,
  selectedNodes: PropTypes.array,
  sortFields: PropTypes.array,
};

ListSelect.defaultProps = {
  details: () => {},
  initialSortOrder: '',
  initialSortDirection: '',
  label: () => {},
  labelKey: '',
  nodes: [],
  onRemoveNode: () => {},
  onSubmitNode: () => {},
  selectedNodes: [],
  sortFields: [],
};

export default ListSelect;
