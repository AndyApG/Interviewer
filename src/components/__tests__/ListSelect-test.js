/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import ListSelect from '../ListSelect';

jest.mock('../../utils/CSSVariables');

const nodes = [
  { uid: 'a', attributes: { name: 'a name', age: '22' } },
  { uid: 'b', attributes: { name: 'b name', age: '88' } },
  { uid: 'c', attributes: { name: 'c name', age: '33' } },
];

describe('ListSelect component', () => {
  it('renders ok', () => {
    const component = shallow((
      <ListSelect
        name="foo"
        nodes={nodes}
      />
    ));

    expect(component).toMatchSnapshot();
  });
});
