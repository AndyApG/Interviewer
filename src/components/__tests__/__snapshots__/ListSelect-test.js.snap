// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListSelect component renders ok 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListSelect
    details={[Function]}
    initialSortOrder={
      Array [
        Object {
          "direction": "asc",
          "property": "",
        },
      ]
    }
    label={[Function]}
    name="foo"
    nodes={
      Array [
        Object {
          "attributes": Object {
            "age": "22",
            "name": "a name",
          },
          "uid": "a",
        },
        Object {
          "attributes": Object {
            "age": "88",
            "name": "b name",
          },
          "uid": "b",
        },
        Object {
          "attributes": Object {
            "age": "33",
            "name": "c name",
          },
          "uid": "c",
        },
      ]
    }
    onRemoveNode={[Function]}
    onSubmitNode={[Function]}
    selectedNodes={Array []}
    sortFields={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="list-select__sort"
        >
          <input
            onChange={[Function]}
            placeholder="Filter"
            type="search"
            value=""
          />
        </div>,
        <Scrollable
          details={[Function]}
          label={[Function]}
          nodes={
            Array [
              Object {
                "attributes": Object {
                  "age": "22",
                  "name": "a name",
                },
                "uid": "a",
              },
              Object {
                "attributes": Object {
                  "age": "88",
                  "name": "b name",
                },
                "uid": "b",
              },
              Object {
                "attributes": Object {
                  "age": "33",
                  "name": "c name",
                },
                "uid": "c",
              },
            ]
          }
          onToggleCard={[Function]}
          selected={[Function]}
        />,
      ],
      "className": "list-select",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [],
            <input
              onChange={[Function]}
              placeholder="Filter"
              type="search"
              value=""
            />,
          ],
          "className": "list-select__sort",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Filter",
              "type": "search",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "details": [Function],
          "label": [Function],
          "nodes": Array [
            Object {
              "attributes": Object {
                "age": "22",
                "name": "a name",
              },
              "uid": "a",
            },
            Object {
              "attributes": Object {
                "age": "88",
                "name": "b name",
              },
              "uid": "b",
            },
            Object {
              "attributes": Object {
                "age": "33",
                "name": "c name",
              },
              "uid": "c",
            },
          ],
          "onToggleCard": [Function],
          "selected": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="list-select__sort"
          >
            <input
              onChange={[Function]}
              placeholder="Filter"
              type="search"
              value=""
            />
          </div>,
          <Scrollable
            details={[Function]}
            label={[Function]}
            nodes={
              Array [
                Object {
                  "attributes": Object {
                    "age": "22",
                    "name": "a name",
                  },
                  "uid": "a",
                },
                Object {
                  "attributes": Object {
                    "age": "88",
                    "name": "b name",
                  },
                  "uid": "b",
                },
                Object {
                  "attributes": Object {
                    "age": "33",
                    "name": "c name",
                  },
                  "uid": "c",
                },
              ]
            }
            onToggleCard={[Function]}
            selected={[Function]}
          />,
        ],
        "className": "list-select",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [],
              <input
                onChange={[Function]}
                placeholder="Filter"
                type="search"
                value=""
              />,
            ],
            "className": "list-select__sort",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Filter",
                "type": "search",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "details": [Function],
            "label": [Function],
            "nodes": Array [
              Object {
                "attributes": Object {
                  "age": "22",
                  "name": "a name",
                },
                "uid": "a",
              },
              Object {
                "attributes": Object {
                  "age": "88",
                  "name": "b name",
                },
                "uid": "b",
              },
              Object {
                "attributes": Object {
                  "age": "33",
                  "name": "c name",
                },
                "uid": "c",
              },
            ],
            "onToggleCard": [Function],
            "selected": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
